content/blog/2020-08-19-train-a-sequence-model-with-poutyne/index.en.md

```diff
@@ -0,0 +1,613 @@
+
+        tags = item[1]
+        idx_tags = self._convert_tags_to_idx(tags)
+
+        return address_vector, idx_tags
+
+    def _convert_tags_to_idx(self, tags):
+        idx_tags = []
+        for tag in tags:
+            idx_tags.append(self.tags_set[tag])
+        return idx_tags
+
+dataset_vectorizer = DatasetVectorizer(embedding_vectorizer)
+```  
+
+Now, let's vectorize our dataset.
+
+```python
+dataset_vectorizer.vectorize(train_data)
+dataset_vectorizer.vectorize(valid_data)
+dataset_vectorizer.vectorize(test_data)
+```
+
+```python
+train_data[0]
+```
+
+> Here is a example after the vectorizing process
+
+```
+([array([ 5.30934185e-02, -6.72420338e-02,  8.96735638e-02, -3.26771051e-01,
+         -4.42410737e-01, -1.32014668e-02,  1.50324404e-01, -2.50251926e-02,
+          1.24011010e-01, -1.68488681e-01,  4.80616689e-02, -5.40233105e-02,
+          1.21191796e-02,  3.89859192e-02,  1.25505164e-01, -1.40419468e-01,
+         -2.62053646e-02,  1.01731330e-01, 
+         ...
+         -5.71550950e-02, -3.92134525e-02,  4.85491045e-02,  4.82993454e-01,
+          3.35614271e-02, -5.97143888e-01, -9.82549414e-02,  8.23293403e-02],
+        dtype=float32), #dimension of 300 # first word
+        ... 
+        # word N of the sequence
+        array([-6.28125593e-02, -1.72182580e-03,  1.27990674e-02,  7.47001171e-02,
+        ...
+        3.74843590e-02],
+        dtype=float32)],
+    
+ #second element of the tuple
+ [0, 1, 1, 1, 3, 4, 5, 5] #the ground truth tag)
+```
+
+### DataLoader
+> We use a first trick, ``padding``.
+
+Now, because the addresses are not all of the same size, it is impossible to batch them together; recall that all tensor elements must have the same lengths. But there is a trick, padding!
+
+The idea is simple; we add *empty* tokens at the end of each sequence until they reach the length of the longest one in the batch. For example, if we have three sequences of length ${1, 3, 5}$, padding will add *empty* tokens to the first two, 4 for the first and 2 for the second.
+
+For the word vectors, we add vectors of 0 as padding. For the tag indices, we pad with -100's. We do so because the [cross-entropy loss](https://pytorch.org/docs/stable/generated/torch.nn.CrossEntropyLoss.html#torch.nn.CrossEntropyLoss) and the accuracy metric both ignore targets with values of -100.
+
+To do the padding, we use the `collate_fn` argument of the [PyTorch `DataLoader`](https://pytorch.org/docs/stable/data.html#torch.utils.data.DataLoader), and on running time, the process will be done by the `DataLoader`. One thing to keep in mind when treating padded sequences is that their (original) length will be required to unpack them later on (since we also pack them), in the forward pass. That way, we can pad and pack the sequence to minimize the training time (read [this good explanation](https://stackoverflow.com/questions/51030782/why-do-we-pack-the-sequences-in-pytorch) of why we pack sequences).
```

Comment:
```suggestion
To do the padding, we use the `collate_fn` argument of the [PyTorch `DataLoader`](https://pytorch.org/docs/stable/data.html#torch.utils.data.DataLoader), and on running time, the process will be done by the `DataLoader`. One thing to keep in mind when treating padded sequences is that their original length will be required to unpad them later on in the forward pass. That way, we can pad and pack the sequence to minimize the training time (read [this good explanation](https://stackoverflow.com/questions/51030782/why-do-we-pack-the-sequences-in-pytorch) on why we pack sequences).
```
Typos.
Also, I don't know why we would speak of unpacking here: the concept of packing here is yet to be talked about and we only need the original lengths for the unpadding if I am correct (might be wrong tbh) ?